### PR TITLE
Static imports

### PR DESCRIPTION
**Pull request by kureta**
Rewrote __init__.py files

### CODE REVIEW

**Comment by kureta** [comment]
@josiah-wolf-oberholtzer which one would you prefer:

```
from abjad.tools import datastructuretools

some_method(datastructuretools.NegativeInfinity)
```

or

```
from abjad.tools.datastructuretools import NegativeInfinity, Infinity, ...

some_method(NegativeInfinity)
```

**Comment by kureta** [comment]
Second option made more sense and it is cleaner. So I pushed the changes.

**Comment by josiah-wolf-oberholtzer** [comment]
I would prefer that those constant object be defined in the root Abjad initializer, not in the `datastructuretools` initializer.

Client code is more likely to reference `Infinity` via `abjad.Infinity`, not `abjad.tools.datastructuretools.Infinity`, or `from abjad import Infinity` instead of `from abjad.tools.datastructuretools import Infinity`.

**Comment by kureta** [comment]
Both `abjad.Infinity` and `from abjad import Infinity` work this way, because we are importing everything in `abjad/__init__.py`. Also, moving these constant definitions into `abjad/__init__.py` is not a trivial task. These constants are used heavily within datastructuretools module and everything within this module is also imported from `tools/__init__.py`. Moving the constant definitions to the top of `abjad/__init__.py` does not solve the issue either. I think moving these into `abjad/__init__.py` is part of a new, bigger issue about library structure.

```
File "/Users/kureta/Documents/sources/abjad/conftest.py", line 1, in <module>
    import abjad
  File "/Users/kureta/Documents/sources/abjad/abjad/__init__.py", line 26, in <module>
    from abjad.tools.datastructuretools.OrdinalConstant import OrdinalConstant
  File "/Users/kureta/Documents/sources/abjad/abjad/tools/__init__.py", line 6, in <module>
    from . import datastructuretools
  File "/Users/kureta/Documents/sources/abjad/abjad/tools/datastructuretools/__init__.py", line 16, in <module>
    from .Sequence import Sequence
  File "/Users/kureta/Documents/sources/abjad/abjad/tools/datastructuretools/Sequence.py", line 10, in <module>
    from abjad import Exact
ImportError: cannot import name 'Exact'
```

Just this line: `from abjad.tools.datastructuretools.OrdinalConstant import OrdinalConstant`, which is required to define the constants to begin with, causes everything to be imported like a chain reaction and prevents us from defining the constants within `abjad/__init__.py`

I think we should open a new issue about package structure. There are problems with imports/late-imports, also having everything under `tools` submodule, and also having `-tools` suffix at the and of every submodule of `tools` is a bit redundant. I don't know... I'll be waiting for a response from you.

**Comment by josiah-wolf-oberholtzer** [comment]
I'll accept a compromise of reimporting the constants back into the root initializer from the datastructuretools module.

**Comment by josiah-wolf-oberholtzer** [comment]
If you want to propose a new package structure, do it!

It's true that `tools` appears twice in most Abjad import paths. If we were to remove anything, it would likely be the intermediate `tools` directory, preserving the `tools` suffix in the package names because the suffix prevents the noun prefix from conflicting with likely variable names a user might actually use.

I'm curious what you'd want to do though.

**Comment by kureta** [comment]
Currently the constants are defined in `datastructuretools/__init__.py` and they are imported into `abjad/__init__.py` via the `from abjad.tools.datastructuretools import *` statement. Is this OK?

I'll take a look into the interdependencies between different modules and come up with a package structure proposal.

**Comment by josiah-wolf-oberholtzer** [comment]
Yes, that works for me! I'll give this PR one more read-through and then merge it, probably this coming week.

**Comment by trevorbaca** [comment]
Hey @josiah-wolf-oberholtzer lemme know how the pull goes. I went ahead and removed `ImportManager` from my own score packages while @kureta was making these changes; curious to try the updated code against all the scores.

Also, I've enjoyed following the discussion about package structure. Package structured has bedeviled my own thinking about Abjad for years, and I'm always open to tightening everything. We just have to be very very careful (and proceed *extremely* linearly) with changes to the library structure as a whole.

I think a good top-level design goal for library structure is that -- as much as possible -- we want composers to just reference all the classes in the system with "Abjad-global" namespacing: `abjad.Articulation`, `abjad.PitchClasssSet`, `abjad.Violin`, meaning that composers should be free to almost completely ignore the subpackages (`indicatortools`, `pitchtools`, `instrumenttools`) in which these classes are defined.

Having said that, though, I've gotten tired of the `-tools` suffixes, too: it seems like noise to see `abjad.instrumenttools.Violin` instead of just `abjad.instruments.Violin`. (For what it's worth, the `-tools` suffixes were previously quite helpful when composers were not Abjad-namespacing everything: most score sourcefiles will have variable names like `instruments` and `pitch`, and these identifiers previously would have conflicted with, say, Abjad packages named `instruments` and `pitch`; but this is no longer a problem now that composers should always reference all Abjad content like `abjad.instruments` and `abjad.pitch`.)

**Comment by trevorbaca** [comment]
I pulled @kureta's static import changes, merged with my dev branch, tested all local code and rebuilt all of my scores: everything passes and rebuilds.

**Comment by josiah-wolf-oberholtzer** [comment]
@trevorbaca Good to hear.

I asked for the ImportManager to be re-instated because I need time for my own tooling to catch up. Junking the ImportManager means either forcing myself to edit initializers every time I create or rename materials or segments, or teaching my tooling how to perform the initializer edits automatically.

Doable, but there's been so many changes recently that I'd rather see one chess piece moved at a time for my own sanity. Once I get those changes made (or perhaps move dynamic import logic directly into my own boilerplate), I'm happy to see the ImportManager retired.

**Comment by trevorbaca** [comment]
Absolutely, @josiah-wolf-oberholtzer; right there with you.

FWIW, I went the route of rewriting all of my score package initializers with by-hand declarations of contents; I've gone rounds with the IDE for a couple of years trying to teach the thing to handle package contents changes automatically, and my conclusion from all that is that it's not worth the effort: quite easy to maintain inits by hand. So (maybe?) you might reach a similar conclusion for `commandlinetools` and the rest of your tooling infrastructure. Or maybe not! ;)

**Comment by josiah-wolf-oberholtzer** [comment]
@trevorbaca I don't think it'll be hard to teach the scripts in `commandlinetools` to handle the initializer edits. I just need to get my scores to run with Abjad 2.21+ in the first place before it's worth overhauling the tooling.

**Comment by kureta** [comment]
It is surprising to see that both of you want everything to be accessible in `abjad.Everything`. I didn't use abjad for any serious work yet, so if both @josiah-wolf-oberholtzer and @trevorbaca think that it is a good idea I don't have much to say. However, I like the structure of [librosa](https://github.com/librosa/librosa). Maybe you can take a look at that when you have time. Having everything under a single root without any hierarchy just gives me a bad feeling. And there are too many `...tools` and it makes me feel like they need to be under 4-5 submodules (like librosa's core, features, utils).

That being said, if you guys insist on having everything accessible from `abjad.` then we do not need to make any radical changes to the package structure. I just need to clean up the way things  are imported here and there. Nevertheless, I think we can come up with better names that will not conflict with anything else, and drop the `...tools` suffix, and hopefully the `tools` directory. Neither of them add any information.

I will be studying the existing package structure this week. Please let me know if you have any ideas about package structure, module names, or anything related. Also @josiah-wolf-oberholtzer let me know if anything comes to your mind as _legacy_ or no longer necessary (like the ImportManager class), so that I can take that into consideration.

**Comment by josiah-wolf-oberholtzer** [comment]
@kureta We've been teaching Abjad for a few years now in week-long intensive courses, and a flattened namespace for the vast majority of common classes and functions makes pedagogy simpler. Think of how `numpy` appears when you call `dir()` on it. Most everything you need to start working is available there immediately. Additionally, referencing classes via the flattened namespace makes score more resilient against our continued refactoring. Users don't need to care if `Note` is housed in `scoretools` or `componenttools` or `core` or whatever.

However, not _everything_ is in the flattened namespace. Many of the "utility" classes found in systemtools, abjadbooktools, commandlinetools, ipythontools etc. are not part of the root Abjad namespace. The contents of some specialized composition / analysis packages are also not part of the root Abjad namespace, such as rhythmmakertools, parts of metertools, and tonalanalysistools.

We're working on extracting some of these subpackages out as entirely separate and optional extension packages (e.g. https://github.com/abjad/abjad-ext-ipython, https://github.com/Abjad/abjad-ext-api, https://github.com/Abjad/abjad-ext-book). `rhythmmakertools` will probably also be extracted, and most or all of `commandlinetools` too. The `Manage*Script` classes found in `commandlinetools` are part of my personal project-management tooling, vaguely analogous to @trevorbaca's https://github.com/Abjad/ide, and don't really need to be part of the main distribution.

Now, all of this said, I'm genuinely curious about your suggestion of a few subpackages, in the spirit of `librosa`. Many many years ago, Abjad had a structure maybe a little more like that. Would you be willing to just write out a list of what classes/functions you would put into what subpackages? Doesn't need to be a full PR or comprehensive.

**Comment by kureta** [comment]
@josiah-wolf-oberholtzer I spent the last week sick in bed so I did not have time to go deep into the existing package structure but I have to say a few things. 

> Users don't need to care if `Note` is housed in `scoretools` or `componenttools` or `core` or whatever.

This might seem like an advantage, especially given that you have intimate knowledge of the contents of Abjad. Also for a beginner user of the library, this might actually be an advantage, but given the size of Abjad, this soon becomes a disadvantage. (A side note: we can introduce a namespace, for example `core`, with an `__init__.py` file similar to current `abjad/__init__.py` file, maybe a bit more simplified, but have an organized module structure under the hood).

In the "Core Concepts" section of the documentation, and in your dissertation, the most important thing Abjad brings to the table, at least from my point of view, is the proposed data structure for a composition. A tree structure extended with spanners and indicators. (All the names I will use from now on can be considered placeholders, we can think of better names if we agree on the structure.)

So my first intuition is to have a `buildingblocks` module which contains `containers`, `leaves`, `spanners`, and `indicators` modules, which in turn contain all container, leaf, spanner, and indicator classes. A `tools` module that contains higher level tools and abstractions which use the previous building blocks. For example contents of the current `rhythmmakertools` would go somewhere here. A `utilities` module should contain stuff related to copying, comparing, or modifying the objects in a score. For example, `quantizationtools` might be a part of this module. An `engraving` module should contain tools for *cosmetic* modifications to a score, such as adding space between certain elements, etc.

Other than these, stuff like `documentationtools`, `commandlinetools`, `abjadbooktools` may stay in their own modules. This structure will be helpful also when/if you decide to convert one of these modules into an extension.

I agree that we have to be extremely carefull before making such huge changes to the structure. No need to rush things. I will start clening up "late imports" and things like that. I am guessing such a clean-up would help us in the long run, if/when we decide to change the package structure.

Also, I cannot stress enough how helpful tools such as PyCharm, or vim plugins, emacs plugins that provide similar functionality are, especially when it comes to a library as big as Abjad. It helped me a lot when trying to learn tensorflow. An organized package structure amplifies this advantage. When you invoke auto-completion on `abjad.` such tools are unable to provide any help/context for you. Now that we have introduced static imports it just dumps almost everything Abjad has to offer, which currently is an improvement. If we can come up with a decent structure, users will be able to browse the code under categories that make sense, and a huge barrier to entry will be removed. At least that is how I approach new libraries, new languages, etc.

Finally, I am sure you are already considering this but, we should remove everything that is not part of Abjad, but part of your personal toolkit. As I said, I did not have time to go through all the components of Abjad but this is the structure I would expect after using Abjad, reading the "Core Concepts", and skimming through your dissertation.

**Comment by josiah-wolf-oberholtzer** [comment]
No worries about your timeline. I just moved across the US this week, and now it's time for holidays, so I won't finish a review until next week. As you say, no need to rush.

I'm happy to wait for you to prepare a sketch of what classes and functions you would expect where. I want you to do this because I think you'll find it's not so clear-cut, and doesn't necessarily make more or less sense than the current regime. There are a tremendous number of classes involved, even when extracting many of our infrastructure packages out. Better not to talk about hypotheticals.

I'm not really sold on code-completion as a guiding principle for package design here. I don't want to hinder it, but I also don't see it as reason to structure our package one way or another. `tensorflow` has over a hundred more names available at top-level than `abjad`. Other major packages like `numpy`, `scipy` and `pandas` also have many more names than `abjad`, in addition to many nested submodules. I've never used code-completion, and know few developers at my company who do. Clear documentation, especially in a dynamic language like Python, has always been more helpful for me, along with simply reading the source code.

That said, let's still talk about reorganizing things :)! I've opened an issue where we can begin to sketch out in text what Abjad 3.0 _could_ look like, and can discuss the pros and cons of each organization: #898

File: abjad/tools/documentationtools/DocumentationManager.py [hunk 1]
```diff
@@ -520,7 +520,7 @@ def _get_tools_package_contents(self, tools_package):
                     pass
                 else:
                     message = 'Warning: {} does not implement a {name} object.'
-                    message = message.format(tools_package, name)
+                    message = message.format(tools_package, name=name)
```
Comment (josiah-wolf-oberholtzer) [inline abjad/tools/documentationtools/DocumentationManager.py]: 🍪

**Pull request merged**